### ISSUE TITLE
Incorrect error message on `external` arity mismatch with local substitution

### ISSUE THREAD

**Issue by craigfe** [ocaml/ocaml#10427]
I ran into the following example today:

```ocaml
module Identity : sig
  type 'a t = 'a
  type ('a, 'b) arrow := 'a -> 'b
  external map : ('a -> 'b) -> ('a t, 'b t) arrow = "%identity"
end = struct
  type 'a t = 'a
  external map : ('a -> 'b) -> 'a t -> 'b t = "%identity"
end
```

This is rightly rejected because the two `external`s have differing arities. (It also happens that the implementation expects the wrong arity of `%identity`, but this won't be determined until after type checking.)

However, the reason that OCaml gives for rejecting it is not correct:

```text
Lines 5-8, characters 6-3:
Error: Signature mismatch:
       ...
       Values do not match:
         external map : ('a -> 'b) -> 'a -> 'b = "%identity"
       is not included in
         external map : ('a -> 'b) -> 'a -> 'b = "%identity"
```

The compiler is making a distinction between two different classes of arrows that isn't conveyed to the user. It seems better for the compiler to state that the arities of the two types differ, and perhaps even refer to the corresponding section of the manual since this notion of arity can be surprising for users (e.g. [here](https://github.com/ocaml/ocaml/issues/8660)).

**Comment by lpw25** [comment]
I think this is fixed by #10407

**Comment by antalsz** [comment]
Yes – the new error will be

<pre><code>Error: Signature mismatch:
       ...
       Values do not match:
         external map : ('a -> 'b) -> 'a t -> 'b t = "%identity"
       is not included in
         external map : ('a -> 'b) -> 'a t -> 'b t = "%identity"
<strong>       The syntactic arities of these primitives were not the same
       (They must have the same number of arrows present in the source)</strong></code></pre>

There is now extra explanatory text that points to the syntactic restriction that's present.  However, because of the destructive substitution, the reported *types* are still unclear.
